### ISSUE TITLE
executable file not found

### ISSUE THREAD

**Issue by kcardona86** [XeroAPI/xoauth#5]
on Linux after using `./xaouth setup` the below error was thrown when attempting to create a token.

`dbus-launch executable file not found in $path `

**Comment by kcardona86** [comment]
`./xaouth delete` gives same error

![image](https://user-images.githubusercontent.com/4367346/80631704-7308d280-8a56-11ea-8284-27d83c9671c2.png)

**Comment by kcardona86** [comment]
./xaouth setup update-secret also 
![image](https://user-images.githubusercontent.com/4367346/80632334-6b95f900-8a57-11ea-9fcb-e13dfb2d5e96.png)

**Comment by kcardona86** [comment]
Cannot delete clientName, as it throws same similar error

**Comment by madhavvishnubhatta** [comment]
I am also trying to experiment with xero and its API, and I ran into the same issue. I installed dbus-x11 which resolved that issue.

```bash
yum install dbus-x11
```

But then a new error popped up.

> error creating client: The name org.freedesktop.secrets was not provided by any .service files

Then I installed gnome-keyring.

```bash
yum install gnome-keyring
```

With that I now have this isue: 
> error creating client: failed to unlock correct collection '/org/freedesktop/secrets/aliases/default'

I am not sure how to proceed further. I suspect this is because we are trying to do it on a Linux machine without a KDE or some such GUI environment. That is just a guess and I have no solid basis to say that.

I am not sure when I will get back to this, but thought I can share my progress so far for what it is worth.

**Comment by kcardona86** [comment]
Don't know if this helps [](https://rtfm.co.ua/en/what-is-linux-keyring-gnome-keyring-secret-service-and-d-bus/)

**Comment by scottjs** [comment]
I'm facing the same issue, this is going to be running on a headless ubuntu server as part of a queued batch process. 

I need the ability to authenticate to this API without having to programatically open a GUI browser to authorise the request, which seems utterly ridiculous. 

Has anyone managed to make any progress with it?

**Comment by YourZero** [comment]
Having the same issue here. I cannot do machine to machine api calls, all because I've opted to use a headless os. Seems like a design flaw on Xero's behalf here. Can you guys please look into this?

**Comment by scottjs** [comment]
> Having the same issue here. I cannot do machine to machine api calls, all because I've opted to use a headless os. Seems like a design flaw on Xero's behalf here. Can you guys please look into this?

It's been a nightmare for us and ended up costing our client much more work than anticipated.

We have a system where when an invoice is raised in our system, we copy it into Xero.

What we ended up having to do was implement a "Xero Authorize" settings page within the client's back office admin panel where an admin user can click the "Authorize" button, then we store the access token and refresh token in the caches. 

Theoretically this only has to be done once.

The refresh token lasts 60 days, so we have a scheduled task that runs every month or so to refresh the token and store the new values in the cache. 

This essentially keeps the tokens alive indefinitely and any new raised invoices will be seamlessly copied over to Xero. 

Not a particularly enjoyable solution.

**Comment by plopidou** [comment]
Just ran into a similar issue, except I get the error below even before.

```
./xoauth setup xero
? What's the Authority? https://identity.xero.com
? What's your client_id? XXXXXXXXXXXXXXXXXXXXXXX
? Select Grant Type: authorization_code
? What's your client_secret? ************************************************
Enter scopes (type `d` to finish)
Scopes are [openid offline_access]
? Add scope (`d` when done) d
error creating client: exec: "dbus-launch": executable file not found in $PATH
```

So, as of today:

- oauth2 client_credentials do not work (design choice, etc, etc.). Fine (almost, see the various threads on the subject on the Xero official forums)
- running xoauth on the command line via ssh breaks as well. Not fine.

How am I/we supposed to handle machine to machine interaction with Xero? Should I book an appointment at the Data center which is 350 miles away and run stuff from there? After of course remotely installing GUIs on machines that are not supposed to run them at all (they are celery/rabbitmq) nodes....? So I can run a `startx` only for Xero? Really?

The documentation mentions "Once the user has given consent, all the tokens will be displayed in the web browser as well as being piped to stdout so you can use them in a script workflow." That's great! How about you test for the existence of a GUI and if not found, simply _only_ pipe?

Also, the page https://developer.xero.com/documentation/api-guides/machine-2-machine has a animated GIF as a documentation. Hard to follow, to say the least.


I think the above is problematic, and Xero is the _only_ third party service that causes such issues (after integrating with Visa, Mastercard, Stripe, Companies House UK, Capsule CRM to name a few).

If I did not follow the steps in a appropriate way, please do point me to the right one


$ ./xoauth doctor
✅ Looking good

Mmmmm
